src/effects/VST3/VST3UIValidator.cpp

```diff
@@ -0,0 +1,205 @@
+
+      ViewRect defaultSize;
+      if(view->getSize(&defaultSize) == kResultOk)
+      {
+         if(view->canResize() == Steinberg::kResultTrue)
+         {
+            ViewRect minSize {0, 0, parent->GetMinWidth(), parent->GetMinHeight()};
+            ViewRect maxSize {0, 0, 10000, 10000};
+            if(view->checkSizeConstraint(&minSize) != kResultOk)
+               minSize = defaultSize;
+
+            if(view->checkSizeConstraint(&maxSize) != kResultOk)
+               maxSize = defaultSize;
+
+            //No need to accommodate the off-by-one error with Steinberg::ViewRect
+            //as we do it with wxRect
+
+            if(defaultSize.getWidth() < minSize.getWidth())
+               defaultSize.right = defaultSize.left + minSize.getWidth();
+            if(defaultSize.getWidth() > maxSize.getWidth())
+               defaultSize.right = defaultSize.left + maxSize.getWidth();
+
+            if(defaultSize.getHeight() < minSize.getHeight())
+               defaultSize.bottom = defaultSize.top + minSize.getHeight();
+            if(defaultSize.getHeight() > maxSize.getHeight())
+               defaultSize.bottom = defaultSize.top + maxSize.getHeight();
+
+            parent->SetMinSize({minSize.getWidth(), minSize.getHeight()});
+            parent->SetMaxSize({maxSize.getWidth(), maxSize.getHeight()});
+         }
+         else
+         {
+            parent->SetMinSize({defaultSize.getWidth(), defaultSize.getHeight()});
+            parent->SetMaxSize(parent->GetMinSize());
+         }
+
+         parent->SetSize({defaultSize.getWidth(), defaultSize.getHeight()});
+      }
+
+#if __WXGTK__
+      mPlugView = view;
+      safenew internal::x11::SocketWindow(parent, wxID_ANY, view);
+      return true;
+#else
+
+      static const auto platformType =
+#  if __WXMAC__
+         kPlatformTypeNSView;
+#  elif __WXMSW__
+         kPlatformTypeHWND;
+#  else
+#     error "Platform not supported"
+#  endif
+      auto plugFrame = owned(safenew internal::PlugFrame { parent });
+      view->setFrame(plugFrame);
+      if(view->attached(parent->GetHandle(), platformType) != kResultOk)
+         return false;
+
+      mPlugView = view;
+      mPlugFrame = plugFrame;
+
+      return true;
+#endif
+   }
+   return false;
+}
+
+bool VST3UIValidator::IsGraphicalUI()
+{
+   return mPlugView != nullptr;
+}
+
+bool VST3UIValidator::ValidateUI()
+{
+   mAccess.ModifySettings([&](EffectSettings &settings){
+      if (mDuration != nullptr)
+         settings.extra.SetDuration(mDuration->GetValue());
+      mWrapper.FlushSettings(settings);
+      mWrapper.StoreSettings(settings);
+   });
+
+   return true;
+}
+
+void VST3UIValidator::OnClose()
+{
+   using namespace Steinberg;
+
+   mPlainUI = nullptr;
+   mParent = nullptr;
+   if(mPlugView)
+   {
+      mPlugView->setFrame(nullptr);
+      mPlugView->removed();
+      mPlugView = nullptr;
+      mPlugFrame = nullptr;
+   }
+}
```

Comment:
I think this should call through to `EffectUIValidator::OnClose` so the effect doesn't hold on to `mParent` either

Reply:
Thanks